File: e2e/php/tests/ContentTest.php
<?php

declare(strict_types=1);

namespace Kreuzberg\E2e;

use PHPUnit\Framework\TestCase;
use Kreuzcrawl\Kreuzcrawl;

/** E2e tests for category: content. */
final class ContentTest extends TestCase
{
    /** Handles 204 No Content response gracefully */
    public function test_content_204_no_content(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals(204, $result->status_code);
        $this->assertEmpty($result->html);
    }

    /** Handles ISO-8859-1 encoded page correctly */
    public function test_content_charset_iso8859(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals("iso-8859-1", $result->content->detected_charset);
    }

    /** Handles 200 response with empty body gracefully */
    public function test_content_empty_body(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals(200, $result->status_code);
    }

    /** Handles response with Accept-Encoding gzip negotiation */
    public function test_content_gzip_compressed(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertNotEmpty($result->html);
        $this->assertEquals(200, $result->status_code);
    }

    /** Respects max body size limit and truncates or skips oversized pages */
    public function test_content_large_page_limit(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertLessThan(1025, $result->content->body_size);
    }

    /** Extracts only main content area, excluding nav, sidebar, footer */
    public function test_content_main_only(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals(true, $result->content->main_content_only);
    }

    /** Detects PDF content by Content-Type header when URL has no .pdf extension */
    public function test_content_pdf_no_extension(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals(true, $result->content->is_pdf);
    }

    /** Removes specified HTML elements by CSS selector before processing */
    public function test_content_remove_tags(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertNotEmpty($result->html);
    }

    /** Handles UTF-8 content with BOM marker correctly */
    public function test_content_utf8_bom(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals("utf-8", $result->content->detected_charset);
        $this->assertNotEmpty($result->html);
    }
}

File: e2e/rust/tests/content_test.rs
//! E2e tests for category: content

use kreuzcrawl::scrape;

#[test]
fn test_content_204_no_content() {
    // Handles 204 No Content response gracefully
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert_eq!(result.status_code, "204", "equals assertion failed");
    assert!(result.html.is_empty(), "expected empty value");
}

#[test]
fn test_content_charset_iso8859() {
    // Handles ISO-8859-1 encoded page correctly
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert_eq!(result.content.detected_charset, r#"iso-8859-1"#, "equals assertion failed");
}

#[test]
fn test_content_empty_body() {
    // Handles 200 response with empty body gracefully
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert_eq!(result.status_code, "200", "equals assertion failed");
}

#[test]
fn test_content_gzip_compressed() {
    // Handles response with Accept-Encoding gzip negotiation
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert!(!result.html.is_empty(), "expected non-empty value");
    assert_eq!(result.status_code, "200", "equals assertion failed");
}

#[test]
fn test_content_large_page_limit() {
    // Respects max body size limit and truncates or skips oversized pages
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert!(result.content.body_size < 1025_f64, "expected < 1025");
}

#[test]
fn test_content_main_only() {
    // Extracts only main content area, excluding nav, sidebar, footer
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert_eq!(result.content.main_content_only, "true", "equals assertion failed");
}

#[test]
fn test_content_pdf_no_extension() {
    // Detects PDF content by Content-Type header when URL has no .pdf extension
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert_eq!(result.content.is_pdf, "true", "equals assertion failed");
}

#[test]
fn test_content_remove_tags() {
    // Removes specified HTML elements by CSS selector before processing
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert!(!result.html.is_empty(), "expected non-empty value");
}

#[test]
fn test_content_utf8_bom() {
    // Handles UTF-8 content with BOM marker correctly
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert_eq!(result.content.detected_charset, r#"utf-8"#, "equals assertion failed");
    assert!(!result.html.is_empty(), "expected non-empty value");
}

